File: backend/app/repositories/product_repository.py
from ..config.database import get_db

class ProductRepository:
    @staticmethod
    def find_all() -> list:
        """
        Fetch all products from PostgreSQL.
        """
        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT id, barcode, name, price, weight, image, category, created_at "
                    "FROM products ORDER BY created_at DESC;"
                )
                columns = [col[0] for col in cur.description]
                rows = cur.fetchall()
                return [dict(zip(columns, row)) for row in rows]

    @staticmethod
    def find_by_id(product_id: str) -> dict | None:
        """
        Fetch a product by UUID.
        """
        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT id, barcode, name, price, weight, image, category, created_at "
                    "FROM products WHERE id = %s;",
                    (product_id,)
                )
                columns = [col[0] for col in cur.description]
                row = cur.fetchone()
                if not row:
                    return None
                return dict(zip(columns, row))

    @staticmethod
    def find_by_barcode(barcode: str) -> dict | None:
        """
        Fetch a product by barcode.
        """
        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT id, barcode, name, price, weight, image, category, created_at "
                    "FROM products WHERE barcode = %s;",
                    (barcode,)
                )
                columns = [col[0] for col in cur.description]
                row = cur.fetchone()
                if not row:
                    return None
                return dict(zip(columns, row))

    @staticmethod
    def create(data: dict) -> dict:
        """
        Insert a new product into PostgreSQL.
        """
        barcode = data["barcode"]
        name = data["name"]
        price = data["price"]
        weight = data.get("weight")
        image = data.get("image")
        category = data.get("category")

        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    INSERT INTO products (barcode, name, price, weight, image, category)
                    VALUES (%s, %s, %s, %s, %s, %s)
                    RETURNING id, barcode, name, price, weight, image, category, created_at;
                    """,
                    (barcode, name, price, weight, image, category)
                )
                columns = [col[0] for col in cur.description]
                row = cur.fetchone()
                conn.commit()
                return dict(zip(columns, row))

    @staticmethod
    def update(product_id: str, data: dict) -> dict | None:
        """
        Update an existing product.
        """
        fields = []
        values = []
        updatable = ["barcode", "name", "price", "weight", "image", "category"]
        for key in updatable:
            if key in data and data[key] is not None:
                fields.append(f"{key} = %s")
                values.append(data[key])
        
        if not fields:
            return ProductRepository.find_by_id(product_id)

        values.append(product_id)
        query = (
            f"UPDATE products SET {', '.join(fields)} WHERE id = %s "
            f"RETURNING id, barcode, name, price, weight, image, category, created_at;"
        )
        
        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute(query, values)
                columns = [col[0] for col in cur.description]
                row = cur.fetchone()
                conn.commit()
                if not row:
                    return None
                return dict(zip(columns, row))

    @staticmethod
    def delete(product_id: str) -> bool:
        """
        Delete a product.
        """
        with get_db() as conn:
            with conn.cursor() as cur:
                cur.execute("DELETE FROM products WHERE id = %s;", (product_id,))
                row_count = cur.rowcount
                conn.commit()
                return row_count > 0

File: backend/src/repositories/product.repository.ts
import { pool } from '../config/database';
import { Product } from '../services/product.service';

export class ProductRepository {
  /**
   * Fetch all products from PostgreSQL database.
   */
  static async findAll(): Promise<Product[]> {
    const { rows } = await pool.query('SELECT * FROM products ORDER BY created_at DESC');
    return rows.map(this.mapRowToProduct);
  }

  /**
   * Fetch a single product by UUID.
   */
  static async findById(id: string): Promise<Product | null> {
    const { rows } = await pool.query('SELECT * FROM products WHERE id = $1', [id]);
    if (rows.length === 0) return null;
    return this.mapRowToProduct(rows[0]);
  }

  /**
   * Fetch a single product by Barcode.
   */
  static async findByBarcode(barcode: string): Promise<Product | null> {
    const { rows } = await pool.query('SELECT * FROM products WHERE barcode = $1', [barcode]);
    if (rows.length === 0) return null;
    return this.mapRowToProduct(rows[0]);
  }

  /**
   * Insert a new product into PostgreSQL.
   */
  static async create(data: Partial<Product>): Promise<Product> {
    const { barcode, name, price, weight, image, category } = data;
    const query = `
      INSERT INTO products (barcode, name, price, weight, image, category)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING *
    `;
    const values = [
      barcode,
      name,
      price,
      weight ?? null,
      image ?? null,
      category ?? null
    ];
    const { rows } = await pool.query(query, values);
    return this.mapRowToProduct(rows[0]);
  }

  /**
   * Update a product by UUID.
   */
  static async update(id: string, data: Partial<Product>): Promise<Product | null> {
    const fields: string[] = [];
    const values: any[] = [];
    let queryIndex = 1;

    const updatableFields = ['barcode', 'name', 'price', 'weight', 'image', 'category'];
    for (const [key, val] of Object.entries(data)) {
      if (updatableFields.includes(key)) {
        fields.push(`${key} = $${queryIndex}`);
        values.push(val);
        queryIndex++;
      }
    }

    if (fields.length === 0) {
      return this.findById(id);
    }

    values.push(id);
    const query = `
      UPDATE products
      SET ${fields.join(', ')}
      WHERE id = $${queryIndex}
      RETURNING *
    `;

    const { rows } = await pool.query(query, values);
    if (rows.length === 0) return null;
    return this.mapRowToProduct(rows[0]);
  }

  /**
   * Delete a product by UUID.
   */
  static async delete(id: string): Promise<boolean> {
    const { rowCount } = await pool.query('DELETE FROM products WHERE id = $1', [id]);
    return (rowCount ?? 0) > 0;
  }

  /**
   * Helper to map raw database row results to Product object.
   */
  private static mapRowToProduct(row: any): Product {
    return {
      id: row.id,
      barcode: row.barcode,
      name: row.name,
      price: typeof row.price === 'string' ? parseFloat(row.price) : row.price,
      weight: row.weight ? (typeof row.weight === 'string' ? parseFloat(row.weight) : row.weight) : undefined,
      image: row.image || undefined,
      category: row.category || undefined,
      createdAt: row.created_at instanceof Date ? row.created_at.toISOString() : new Date(row.created_at).toISOString(),
      
      // Compatibility layers with previous mock models
      sku: row.barcode,
      cost: typeof row.price === 'string' ? parseFloat(row.price) * 0.7 : row.price * 0.7,
      stock: 100,
      supplierId: 'sup-1'
    };
  }
}
